scipy/optimize/_constraints.py

```diff
@@ -0,0 +1,620 @@
+
+    return lagr_hess
+
+
+def concatenate_canonical_constraints(constraints, sparse_jacobian=None):
+    """Concatenate sequence of CanonicalConstraint's."""
+    # Compute number of constraints
+    n_eq = 0
+    n_ineq = 0
+    for constr in constraints:
+        n_eq += constr.n_eq
+        n_ineq += constr.n_ineq
+
+    # Concatenate equality constraints
+    def constr_eq(x):
+        constr_eq_list = []
+        for constr in constraints:
+            constr_eq_list += [constr.constr_eq(x)]
+        return np.hstack(constr_eq_list)
+
+    # Concatenate inequality constraints
+    def constr_ineq(x):
+        constr_ineq_list = []
+        for constr in constraints:
+            constr_ineq_list += [constr.constr_ineq(x)]
+        return np.hstack(constr_ineq_list)
+
+    # Concatanate equality constraints Jacobian matrices
+    def jac_eq(x):
+        # Read sequentially all jacobians
+        jac_eq_list = []
+        sparse_jac_list = []
+        for constr in constraints:
+            J = constr.jac_eq(x)
+            jac_eq_list += [J]
+            sparse_jac_list += [spc.issparse(J)]
+        # Use sparse matrix if any of the constraints
+        # are sparse
+        if sparse_jacobian is None:
+            use_sparse = np.any(sparse_jac_list)
+        else:
+            use_sparse = sparse_jacobian
+        # Convert all values to same format
+        # this is done internally anyway and
+        # it helps avoiding some odd behaviours
+        for i in range(len(jac_eq_list)):
+            if use_sparse:
+                jac_eq_list[i] = spc.coo_matrix(jac_eq_list[i])
+            elif sparse_jac_list[i]:
+                jac_eq_list[i] = jac_eq_list[i].toarray()
+        # Concatenate all
+        if use_sparse:
+            return spc.vstack(jac_eq_list, format="csc")
+        else:
+            return np.vstack(jac_eq_list)
+
+    # Concatanate inequality constraints Jacobian matrices
+    def jac_ineq(x):
+        # Read sequentially all jacobians
+        jac_ineq_list = []
+        sparse_jac_list = []
+        for constr in constraints:
+            J = constr.jac_ineq(x)
+            jac_ineq_list += [J]
+            sparse_jac_list += [spc.issparse(J)]
+        # Use sparse matrix if any of the constraints
+        # are sparse in the case sparse_jacobian=None
+        if sparse_jacobian is None:
+            use_sparse = np.any(sparse_jac_list)
+        else:
+            use_sparse = sparse_jacobian
+        # Convert all values to same format
+        # this is done internally anyway and
+        # it helps avoiding some odd behaviours
+        for i in range(len(jac_ineq_list)):
+            if use_sparse:
+                jac_ineq_list[i] = spc.coo_matrix(jac_ineq_list[i])
+            elif sparse_jac_list[i]:
+                jac_ineq_list[i] = jac_ineq_list[i].toarray()
+        # Concatenate all
+        if use_sparse:
+            return spc.vstack(jac_ineq_list, format="csc")
```

Comment:
I will try to do some more detailed analysis later.

So far I have done a few time profiles and this do not seems to degrade the performance significantly.